src/clarity-angular/utils/virtual-scroll/virtual-for-of.ts

```diff
@@ -0,0 +1,139 @@
+/*
+ * Copyright (c) 2016-2017 VMware, Inc. All Rights Reserved.
+ * This software is released under MIT license.
+ * The full license information can be found in LICENSE in the root directory of this project.
+ */
+import {
+    AfterViewInit,
+    ChangeDetectorRef,
+    Directive,
+    EmbeddedViewRef,
+    Input,
+    Renderer2,
+    TemplateRef,
+    ViewContainerRef
+} from "@angular/core";
+
+import {
+    appendAndPreserveScroll,
+    offsetHeight,
+    prependAndPreserveScroll,
+    ratioBottomReady,
+    ratioTopReady
+} from "./dom-helpers";
+import {isNonNgIterable, NonNgIterable} from "./non-ng-iterable";
+
+enum Side {
+    START = -1,
+    END = 1
+}
+
+@Directive({selector: "[clrVirtualForOf]"})
+export class VirtualForOf<T> implements AfterViewInit {
+    constructor(private viewContainer: ViewContainerRef, private template: TemplateRef<{$implicit: T}>,
+                private renderer: Renderer2, private cdr: ChangeDetectorRef) {
+        this.containerEl = <HTMLElement>renderer.parentNode(viewContainer.element.nativeElement);
```

Comment:
Why not just get `ElementRef` and then get the parent from there?
`this.elmentRef.nativeElement.parentElement`?

Reply:
Because your solution uses HTML native elements, whereas this i-one uses the Renderer. As you can see, I'm never going deeper than `elRef.nativeElement`.

It's the same difference as between `elRef.nativeElement. setAttribute("Hello", "World")` and `renderer.setAttribute(elRef.nativeElement, "Hello", "World")`. The second one will work for server-side rendering or other contexts.